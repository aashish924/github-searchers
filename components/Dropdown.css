.custom-select-wrapper {
  position: relative;
  user-select: none;
  width: 128px;
  /* width: 100%; */
  margin: 4px;
}
.custom-select {
  position: relative;
  display: flex;
  width: 100%;
  flex-direction: column;
  border-width: 0 2px 0 2px;
  border-style: solid;
  border-color: #008c99;
}
.custom-select__trigger {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 8px;
  font-size: 16px;
  font-weight: 300;
  color: #3b3b3b;
  height: 48px;
  line-height: 48px;
  background: #ffffff;
  cursor: pointer;
  border-width: 2px 0 2px 0;
  border-style: solid;
  border-color: #008c99;
}
.custom-options {
  position: absolute;
  display: block;
  /* top: 100%; */
  left: 0;
  right: 0;
  border: 2px solid #008c99;
  /* border-top: 0; */
  background: #fff;
  transition: all 0.5s;
  opacity: 0;
  visibility: hidden;
  pointer-events: none;
  z-index: 2;
}
.custom-select.open .custom-options {
  opacity: 1;
  visibility: visible;
  pointer-events: all;
}
.custom-option {
  position: relative;
  display: block;
  padding: 0 8px;
  font-size: 18px;
  font-weight: 300;
  color: #3b3b3b;
  line-height: 48px;
  cursor: pointer;
  transition: all 0.5s;
}
.custom-option:hover {
  cursor: pointer;
  background-color: #fafafa;
}
.custom-option.selected {
  color: #ffffff;
  background-color: #008c99;
}

.arrow {
  position: relative;
  height: 12px;
  width: 12px;
}
.arrow::before, .arrow::after {
  content: "";
  position: absolute;
  bottom: 0px;
  width: 0.14rem;
  height: 100%;
  transition: all 0.5s;
}
.arrow::before {
  left: -5px;
  transform: rotate(45deg);
  background-color: #008c99;
}
.arrow::after {
  left: 5px;
  transform: rotate(-45deg);
  background-color: #008c99;
}
.open .arrow::before {
  left: -5px;
  transform: rotate(-45deg);
}
.open .arrow::after {
  left: 5px;
  transform: rotate(45deg);
}