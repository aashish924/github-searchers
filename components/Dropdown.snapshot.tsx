import * as React from 'react';
import renderer from 'react-test-renderer';

import Dropdown from './Dropdown';

describe('Dropdown', () => {
    it('Dropdownshould render correctly when selected is missing', () => {
        const options = [
            { label: 'Ali', value: 'Ali' },
            { label: 'Ahmed', value: 'Ahmed' },
        ];
        const tree = renderer.create(<Dropdown options={options} />).toJSON();
        expect(tree).toMatchSnapshot('without-selected-option');
    });

    it('Dropdownshould render correctly when selected is provided', () => {
        const options = [
            { label: 'Ali', value: 'Ali' },
            { label: 'Ahmed', value: 'Ahmed' },
        ];
        const tree = renderer.create(<Dropdown options={options} selected="Ali" />).toJSON();
        expect(tree).toMatchSnapshot('with-selected-option');
    });
});
