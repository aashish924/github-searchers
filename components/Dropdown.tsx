import React, { useEffect, useRef, useState } from 'react';

import './Dropdown.css';

export interface Option {
    label: string;
    value: any;
}

export interface SelectProps {
    name?: string;
    options: Option[];
    selected?: string;
    onChange?: (option: Option) => void;
}

/**
 * Hook that alerts clicks outside of the passed ref
 */
const useOutsideAlerter = (ref, callback) => {
    useEffect(() => {
        /**
         * Alert if clicked on outside of element
         */
        function handleClickOutside(event) {
            if (ref.current && !ref.current.contains(event.target)) {
                callback();
            }
        }

        // Bind the event listener
        document.addEventListener('mousedown', handleClickOutside);
        return () => {
            // Unbind the event listener on clean up
            document.removeEventListener('mousedown', handleClickOutside);
        };
    }, [ref]);
};

export const Select = (props: SelectProps) => {
    const [open, setOpen] = useState(false);
    const [selected, setSelected] = useState(props.selected || '');
    const wrapperRef = useRef(null);
    useOutsideAlerter(wrapperRef, () => setOpen(false));

    const selectOption = (option) => {
        setSelected(option.label);
        setOpen(false);
        props.onChange!(option);
    };

    return (
        <div className="custom-select-wrapper" ref={wrapperRef}>
            <div className={open ? 'custom-select open' : 'custom-select'}>
                <div className="custom-select__trigger" onClick={() => setOpen(!open)}>
                    <span>{selected}</span>
                    <div className="arrow"></div>
                </div>
                <div className="custom-options">
                    {props.options.map((option, index) => {
                        return (
                            <span
                                key={`${props.name}-${option.label}-${index}`}
                                className={selected === option.label ? 'custom-option selected' : 'custom-option'}
                                data-value={option.value}
                                onClick={() => selectOption(option)}
                            >
                                {option.label}
                            </span>
                        );
                    })}
                </div>
            </div>
        </div>
    );
};

export default Select;
