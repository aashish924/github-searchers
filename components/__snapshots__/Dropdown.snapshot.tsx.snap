// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Dropdownshould render correctly when selected is missing: without-selected-option 1`] = `
<div
  className="custom-select-wrapper"
>
  <div
    className="custom-select"
  >
    <div
      className="custom-select__trigger"
      onClick={[Function]}
    >
      <span>
        
      </span>
      <div
        className="arrow"
      />
    </div>
    <div
      className="custom-options"
    >
      <span
        className="custom-option"
        data-value="Ali"
        onClick={[Function]}
      >
        Ali
      </span>
      <span
        className="custom-option"
        data-value="Ahmed"
        onClick={[Function]}
      >
        Ahmed
      </span>
    </div>
  </div>
</div>
`;

exports[`Dropdown Dropdownshould render correctly when selected is provided: with-selected-option 1`] = `
<div
  className="custom-select-wrapper"
>
  <div
    className="custom-select"
  >
    <div
      className="custom-select__trigger"
      onClick={[Function]}
    >
      <span>
        Ali
      </span>
      <div
        className="arrow"
      />
    </div>
    <div
      className="custom-options"
    >
      <span
        className="custom-option selected"
        data-value="Ali"
        onClick={[Function]}
      >
        Ali
      </span>
      <span
        className="custom-option"
        data-value="Ahmed"
        onClick={[Function]}
      >
        Ahmed
      </span>
    </div>
  </div>
</div>
`;
